File: src/main/java/aurocosh/divinefavor/common/item/ItemInviteMarker.java
package aurocosh.divinefavor.common.item;

import aurocosh.divinefavor.common.core.DivineFavorCreativeTab;
import aurocosh.divinefavor.common.item.base.ModItem;
import aurocosh.divinefavor.common.util.UtilEntity;
import aurocosh.divinefavor.common.util.UtilNbt;
import aurocosh.divinefavor.common.util.UtilPlayer;
import net.minecraft.client.resources.I18n;
import net.minecraft.client.util.ITooltipFlag;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.entity.player.EntityPlayerMP;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.server.management.PlayerList;
import net.minecraft.util.ActionResult;
import net.minecraft.util.EnumActionResult;
import net.minecraft.util.EnumFacing;
import net.minecraft.util.EnumHand;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;

import javax.annotation.Nullable;
import java.util.List;
import java.util.UUID;

public class ItemInviteMarker extends ModItem {
    public static String TAG_PLAYER_UUID = "PlayerUUID";

    private boolean canTeleportToDimensions;

    public ItemInviteMarker(String name, boolean canTeleportToDimensions) {
        super(name, name);
        this.canTeleportToDimensions = canTeleportToDimensions;
        setMaxStackSize(1);
        setCreativeTab(DivineFavorCreativeTab.INSTANCE);
    }

    @Override
    public ActionResult<ItemStack> onItemRightClick(World world, EntityPlayer player, EnumHand hand) {
        ItemStack stack = player.getHeldItem(hand);
        if(!teleportPlayer(stack, player, world))
            return new ActionResult<>(EnumActionResult.PASS, stack);
        return new ActionResult<>(EnumActionResult.SUCCESS, ItemStack.EMPTY);
    }

    @Override
    public EnumActionResult onItemUse(EntityPlayer player, World world, BlockPos pos, EnumHand hand, EnumFacing facing, float hitX, float hitY, float hitZ) {
        ItemStack stack = player.getHeldItem(hand);
        if(!teleportPlayer(stack, player, world))
            return EnumActionResult.PASS;
        return EnumActionResult.SUCCESS;
    }

    private boolean teleportPlayer(ItemStack stack, EntityPlayer player, World world){
        if(world.isRemote)
            return false;
        if(stack.isEmpty())
            return false;
        if(!(stack.getItem() instanceof ItemInviteMarker))
            return false;
        if(!UtilNbt.checkForTag(stack, TAG_PLAYER_UUID))
            return false;

        NBTTagCompound nbt = UtilNbt.getTag(stack);
        UUID uuid = UUID.fromString(nbt.getString(TAG_PLAYER_UUID));
        PlayerList playerList = world.getMinecraftServer().getPlayerList();
        EntityPlayerMP targetPlayer = playerList.getPlayerByUUID(uuid);
        if (player == null)
            return false;

        if(!canTeleportToDimensions && targetPlayer.dimension != player.dimension)
            return false;
        UtilEntity.teleport(player, targetPlayer.dimension, targetPlayer.getPosition());
        UtilPlayer.damageStack(player, stack);
        return true;
    }

    @Override
    public void addInformation(ItemStack stack, @Nullable World world, List<String> tooltip, ITooltipFlag flag) {
        super.addInformation(stack, world, tooltip, flag);
        NBTTagCompound compound = UtilNbt.getTag(stack);
        if (compound.hasKey(TAG_PLAYER_UUID))
            tooltip.add(I18n.format("item.divinefavor:contract.player", compound.getString(TAG_PLAYER_UUID)));
    }
}

File: src/main/kotlin/aurocosh/divinefavor/common/item/gems/ItemInviteMarker.kt
package aurocosh.divinefavor.common.item.gems

import aurocosh.divinefavor.DivineFavor
import aurocosh.divinefavor.common.constants.ConstGemTabOrder
import aurocosh.divinefavor.common.item.base.ModItem
import aurocosh.divinefavor.common.item.gems.base.IUsableGemItem
import aurocosh.divinefavor.common.item.gems.properties.GemMaskProperties
import aurocosh.divinefavor.common.item.spell_talismans.context.CastContext
import aurocosh.divinefavor.common.item.spell_talismans.context.CastContextGenerator
import aurocosh.divinefavor.common.item.talisman.ISelectedStackProvider
import aurocosh.divinefavor.common.lib.EmptyConst.invalidUUID
import aurocosh.divinefavor.common.lib.extensions.divinePlayerData
import aurocosh.divinefavor.common.lib.extensions.get
import aurocosh.divinefavor.common.lib.extensions.isPropertySet
import aurocosh.divinefavor.common.spirit.base.ModSpirit
import aurocosh.divinefavor.common.stack_properties.StackPropertyHandler
import aurocosh.divinefavor.common.stack_properties.interfaces.IPropertyAccessor
import aurocosh.divinefavor.common.util.UtilEntity
import aurocosh.divinefavor.common.util.UtilItemStack
import net.minecraft.client.resources.I18n
import net.minecraft.client.util.ITooltipFlag
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.ItemStack
import net.minecraft.util.ActionResult
import net.minecraft.util.EnumActionResult
import net.minecraft.util.EnumFacing
import net.minecraft.util.EnumHand
import net.minecraft.util.math.BlockPos
import net.minecraft.world.World

class ItemInviteMarker(name: String, private val canTeleportToDimensions: Boolean, override val consumeOnUse: Boolean, override val favorCost: Int, override val spirit: ModSpirit, maxStackSize: Int) : ModItem(name, name, ConstGemTabOrder.OTHER_GEMS), IUsableGemItem, ISelectedStackProvider {
    override val properties: IPropertyAccessor = propertyHandler

    init {
        setMaxStackSize(maxStackSize)
        creativeTab = DivineFavor.TAB_GEMS
    }

    override fun onItemRightClick(world: World, player: EntityPlayer, hand: EnumHand): ActionResult<ItemStack> {
        val stack = player.getHeldItem(hand)
        val context = CastContextGenerator.rightClick(world, player, hand, stack, stack)
        val result = cast(context)
        return UtilItemStack.actionResult(result, stack)
    }

    override fun onItemUse(player: EntityPlayer, world: World, pos: BlockPos, hand: EnumHand, facing: EnumFacing, hitX: Float, hitY: Float, hitZ: Float): EnumActionResult {
        val stack = player.getHeldItem(hand)
        val context = CastContextGenerator.useCast(player, world, pos, hand, facing, stack, stack)
        val result = cast(context)
        return UtilItemStack.actionResultPass(result)
    }

    override fun cast(context: CastContext): Boolean {
        val stack = context.stack
        if (stack.isEmpty)
            return false
        if (stack.item !is ItemInviteMarker)
            return false
        if (!stack.isPropertySet(ItemWarpMarker.position))
            return false

        val world = context.world
        val uuid = stack.get(playerId)
        val server = world.minecraftServer ?: return false
        val playerList = server.playerList
        val targetPlayer = playerList.getPlayerByUUID(uuid)
        val player = context.player

        val dimension = stack.get(ItemWarpMarker.dimension)
        if (!canTeleportToDimensions && dimension != player.dimension)
            return false
        if (!consumeFavor(player))
            return false
        if (world.isRemote)
            return false

        UtilEntity.teleport(player, dimension, targetPlayer.position)
        if (consumeOnUse)
            stack.shrink(1)
        return true
    }

    private fun consumeFavor(player: EntityPlayer): Boolean {
        if (favorCost == 0)
            return true
        val spiritData = player.divinePlayerData.spiritData
        return spiritData.consumeFavor(spirit.id, favorCost)
    }


    override fun addInformation(stack: ItemStack, world: World?, tooltip: MutableList<String>, flag: ITooltipFlag) {
        super.addInformation(stack, world, tooltip, flag)
        if (stack.isPropertySet(playerName))
            tooltip.add(I18n.format("item.divinefavor:contract.player", stack.get(playerName)))
    }

    override fun getSelectedStack(stack: ItemStack): ItemStack = stack

    companion object {
        val propertyHandler = StackPropertyHandler()
        val maskItemId = propertyHandler.registerProperty(GemMaskProperties.maskItemId)
        val maskItemMeta = propertyHandler.registerProperty(GemMaskProperties.maskItemMeta)
        val playerId = propertyHandler.registerUUIDProperty("PlayerUUID", invalidUUID())
        val playerName = propertyHandler.registerStringProperty("PlayerName", "")
    }
}